File: library/kotlin/src/io/envoyproxy/envoymobile/RequestMapper.kt
package io.envoyproxy.envoymobile

internal fun Request.outboundHeaders(): Map<String, List<String>> {
  val retryPolicyHeaders = retryPolicy?.outboundHeaders() ?: emptyMap()

  val result = mutableMapOf<String, List<String>>()
  result.putAll(headers.filter { entry -> !(entry.key.startsWith(":") || entry.key.startsWith("x-envoy-mobile"))})
  result.putAll(retryPolicyHeaders)
  result[":method"] = listOf(method.stringValue())
  result[":scheme"] = listOf(scheme)
  result[":authority"] = listOf(authority)
  result[":path"] = listOf(path)
  result["x-envoy-mobile-upstream-protocol"] = listOf(upstreamHttpProtocol.stringValue)

  return result
}

private fun RequestMethod.stringValue(): String {
  return when (this) {
    RequestMethod.DELETE -> "DELETE"
    RequestMethod.GET -> "GET"
    RequestMethod.HEAD -> "HEAD"
    RequestMethod.OPTIONS -> "OPTIONS"
    RequestMethod.PATCH -> "PATCH"
    RequestMethod.POST -> "POST"
    RequestMethod.PUT -> "PUT"
    RequestMethod.TRACE -> "TRACE"
  }
}

File: library/swift/src/RequestMapper.swift
private let kRestrictedHeaderPrefix = [":", "x-envoy-mobile"]

extension Request {
  /// Returns a set of outbound headers that include HTTP
  /// information on the URL, method, and additional headers.
  ///
  /// - returns: Outbound headers to send with an HTTP request.
  func outboundHeaders() -> [String: [String]] {
    var headers = self.headers
      .filter { !kRestrictedHeaderPrefix.contains(where: $0.key.hasPrefix) }
      .reduce(into: [
        ":method": [self.method.stringValue],
        ":scheme": [self.scheme],
        ":authority": [self.authority],
        ":path": [self.path],
        "x-envoy-mobile-upstream-protocol": [self.upstreamHttpProtocol.stringValue],
      ]) { $0[$1.key] = $1.value }

    if let retryPolicy = self.retryPolicy {
      headers = headers.merging(retryPolicy.outboundHeaders()) { _, retryHeader in retryHeader }
    }

    return headers
  }
}
